modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js

```diff
@@ -27,16 +27,16 @@ import {
 
 import {BAR_CHART, COLORS} from '../utils/constants';
 
-export default function AuditBarChart({categories, rtl}) {
+export default function AuditBarChart({rtl, vocabularies}) {
 	const auditBarChartData = useMemo(() => {
 		const dataKeys = new Set();
 
-		const bars = categories.reduce((acc, category) => {
-			if (!category.children) {
-				return undefined;
+		const bars = vocabularies.reduce((acc, category) => {
+			if (!category.categories) {
+				return [];
```

Comment:
The point @bryceosterhaus was making [on the other PR](https://github.com/liferay-frontend/liferay-portal/pull/25) is that he would have expected you to return `acc` instead of `undefined`. Imagine this scenario:

1. Loop iteration 1; if we have `categories`, we return `[[{dataKey: 'foo', name: 'foo'}]]`.
2. Loop iteration 2; no `categories`, return `[]`.
3. Loop iteration 3; if we have categories, we return `[[{dataKey: 'bar', name: 'bar'}]]`.

ie. the fact that you had one "bad" vocabulary (the second one) means you reset your array and drop the preceding one(s). Your end result contains only the stuff you returned from the 3rd iteration of the loop.

Now, I suppose that in practice, we always expect these vocabularies to have categories? If that's the case, why do we even need the `if`? But if we *do* expect some of them not to have categories, then it seems you should be returning `acc` here instead of `[]` to avoid the problem of blowing away any prior valid ones.

Reply:
I'm aware of that, thanks for point it out. Maybe it's interesting what it's going to be injected in the `vocabularies` prop via backend:

1. There are no vocabularies: the `EmptyAuditBarChart` is going to be rendered.
```js
const vocabularies = [];
```

2. There are vocabularies with categories: the `AuditBarChart` is going to be rendered with the vocabularies in the X axis and the categories will be displayed grouped in each vocabulary as bars and in the legend (see https://user-images.githubusercontent.com/15845466/84938125-0a94b400-b0dd-11ea-9668-58ed9331d2a2.png):
```js
const vocabularies = [
	{
		categories: [
			{key: 'education', name: 'Education', value: 478, vocabularyName: 'Stage'},
			{key: 'selection', name: 'Selection', value: 1055, vocabularyName: 'Stage'},
			{key: 'selection', name: 'Selection', value: 822, vocabularyName: 'Stage'},
		],
		key: 'business-decision-maker',
		name: 'Business Decision Maker',
		vocabularyName: 'Audience',
	},
	{
		categories: [
			{key: 'education', name: 'Education', value: 125, vocabularyName: 'Stage'},
			{key: 'selection', name: 'Selection', value: 1906, vocabularyName: 'Stage'},
			{key: 'solution', name: 'Solution', value: 987, vocabularyName: 'Stage'},
		],
		key: 'business-end-user',
		name: 'Business End User',
		vocabularyName: 'Audience',
	},
	{
		categories: [
			{key: 'education', name: 'Education', value: 444, vocabularyName: 'Stage'}
			{key: 'selection', name: 'Selection', value: 1733, vocabularyName: 'Stage'}
			{key: 'solution', name: 'Solution', value: 1807, vocabularyName: 'Stage'}
		],
		key: 'technical-decision-maker',
		name: 'Technical Decision Maker',
		vocabularyName: 'Audience',
	},
	{
		categories: [
			{key: 'education', name: 'Education', value: 125, vocabularyName: 'Stage'}
			{key: 'selection', name: 'Selection', value: 317, vocabularyName: 'Stage'}
			{key: 'solution', name: 'Solution', value: 187, vocabularyName: 'Stage'}
		],
		key: 'technical-end-user',
		name: 'Technical End User',
		vocabularyName: 'Audience',
	},
];
```

3. There are vocabularies without categories: the `AuditBarChart` is going to be rendered with the vocabularies in the X axis. Only one bar for each vocabulary with the value and no legend. (see https://user-images.githubusercontent.com/15845466/84938119-08caf080-b0dd-11ea-8c12-8a23f890d62a.png):
```js
const vocabularies = [
	{
		key: 'business-decision-maker',
		name: 'Business Decision Maker',
		value: 987,
		vocabularyName: 'Audience',
	},
	{
		key: 'business-end-user',
		name: 'Business End User',
		value: 1095,
		vocabularyName: 'Audience',
	},
	{
		key: 'technical-decision-maker',
		name: 'Technical Decision Maker',
		value: 2020,
		vocabularyName: 'Audience',
	},
	{
		key: 'technical-end-user',
		name: 'Technical End User',
		value: 422,
		vocabularyName: 'Audience',
	},
];
```

**In summary,** we are not thinking about having vocabularies with categories and other ones without it. If that is the case at any point, we should change `return [];` to `return acc;` to not blowing away any prior valid ones as you said.

Thanks :)

Reply:
Ok, you're right. This is only mock data, no real data. I'll update the PR to change that return 😄

Reply:
Just to clarify... you do realize it's not only mock data vs real data, but that returning `[]` in a reduce is an instant red flag regardless?

Reply:
Yes, we are. Thanks for the feedback. I'll keep it in mind for the future 👍

Reply:
😘